### ISSUE TITLE
the r_square between iou and bbox variance in the refine(jitter(bbox)) method

### ISSUE THREAD

**Issue by Jack-Hu-2001** [microsoft/SoftTeacher#44]
The scatter plot  in your paper about the relationship between iou and the bbox variance(after jittered) is really interesting and showed a strong correlation. Since that, I wonna try another method on single stage detector about estimating the bbox quality under your soft teacher architecture. I simply want to know what's the r_square you've achieved  with soft teacher and faster-rcnn+FPN on COCO 1% labeled dataset. Maybe I wonna have a comparision in my projects in the future. Of course if I could come out with some methods under your architecture, I'll show my greatest gratitude and acknowledgements in my paper or  project report! Sincerely thanks forn your help!!!

**Comment by Jack-Hu-2001** [comment]
If it's not calcuted in your research before(I guess you must have done this, haha), I'll simply run it on my machine and write code about analyzing this metric. Sorry for bothering you with this!
Happy National Day!!!

**Comment by Jack-Hu-2001** [comment]
Sincerely thanks for your help! I would be grateful if I don't need to spend more time getting a result maybe you've already calculated.

**Comment by Jack-Hu-2001** [comment]
Thanks so much for your sincere help!!! Let's discuss maybe we can work it out~

**Comment by Jack-Hu-2001** [comment]
I guess it should be so after carefully checking the function scipy usage here.
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.linregress.html
But previously I expected it to be a little bit higher than this, just according to the scatter plot(just by my eyes, haha).
After all, estimating the quality of bbox_pred  is pretty a hard thing in unlabeled data. 
To confirm, these points(x,y) are gathered from all the proposed bboxes by the teacher in unlabeled images during the whole trainning process, right? If so, this value makes sense in my opinion.
If possible, please let us see the result of only testing bboxes  whose bbox_variance are over the thres set in train_cfg (I guess this is 0.01 from base.py) . I believe this makes more sense because only these bboxes(variance<thr) are used as pseudo bboxes and may produce a better and more reasonable result. As for implementation, just filter those (x,y) pairs whose x is larger than 0.01. I think it would be not hard to do.
Thanks a lot for your effort! Sorry for bothering you with this!

**Comment by MendelXu** [comment]
No. I just evaluate the whole validation set with a trained model and take the result as the target boxes. I didn't try to retrain a model and collect its per epoch/step results. Do you think it is necessary?

**Comment by Jack-Hu-2001** [comment]
Well, as I can see, I believe it's not necessary now and count this on the big COCO dataset during the whole process needs a lot of time and calculation cost. Since my dataset of medical image(5k images, rgb) is pretty small and easy(only 7 classes and a little bit hard for humans to label), I did that today on my single stage detector of retinanet-fpn-GFocalV2(https://github.com/implus/GFocalV2, CVPR2021)( just in the supervised process for 5% labeled data) and the correlation R-square(between iou and top-k in predicted probablility distribution tensor of bbox) reached an average of 0.6 during whole training process and over 0.7 in val set. With this result, I believe this correlation is good enougn for me to carry on my medical projects and I'll try to integrate the gfocal_head into this soft-teacher semi-supervised archetecture. It's pretty probable that my idea could work on my dataset. But I guess it'll still be pretty hard to make it work on COCO dataset. Maybe I could have a try. If this could work, I'll create a PR or get another repo with your link.
Meanwhile, it still needs hard work to estimate the bbox quality on other heads like YOLO-Head, I 'm still trying hard.
Thanks  for your help! YOU DID HELPED ME SO MUCH!

**Comment by MendelXu** [comment]
> Well, as I can see, I believe it's not necessary now and count this on the big COCO dataset during the whole process needs a lot of time and calculation cost. Since my dataset of medical image(5k images, rgb) is pretty small and easy(only 7 classes and a little bit hard for humans to label), I did that today on my single stage detector of retinanet-fpn-GFocalV2(https://github.com/implus/GFocalV2, CVPR2021)( just in the supervised process for 5% labeled data) and the correlation R-square(between iou and top-k in predicted probablility distribution tensor of bbox) reached an average of 0.6 during whole training process and over 0.7 in val set. With this result, I believe this correlation is good enough for me to carry on my medical projects and I'll try to integrate the gfocal_head into this soft-teacher semi-supervised archetecture. It's pretty probable that my idea could work on my dataset. But I guess it'll still be pretty hard to make it work on COCO dataset. Maybe I could have a try. If this could work, I'll create a PR or get another repo with your link. Meanwhile, it still needs hard work to estimate the bbox quality on other heads like YOLO-Head, I 'm still trying hard. Thanks for your help! YOU DID HELPED ME SO MUCH!

Nice job. I think converting the regression problem to a classification problem is a possible solution and it is easier to get the statical characteristic from a distribution.

**Comment by Jack-Hu-2001** [comment]
> > Well, as I can see, I believe it's not necessary now and count this on the big COCO dataset during the whole process needs a lot of time and calculation cost. Since my dataset of medical image(5k images, rgb) is pretty small and easy(only 7 classes and a little bit hard for humans to label), I did that today on my single stage detector of retinanet-fpn-GFocalV2(https://github.com/implus/GFocalV2, CVPR2021)( just in the supervised process for 5% labeled data) and the correlation R-square(between iou and top-k in predicted probablility distribution tensor of bbox) reached an average of 0.6 during whole training process and over 0.7 in val set. With this result, I believe this correlation is good enough for me to carry on my medical projects and I'll try to integrate the gfocal_head into this soft-teacher semi-supervised archetecture. It's pretty probable that my idea could work on my dataset. But I guess it'll still be pretty hard to make it work on COCO dataset. Maybe I could have a try. If this could work, I'll create a PR or get another repo with your link. Meanwhile, it still needs hard work to estimate the bbox quality on other heads like YOLO-Head, I 'm still trying hard. Thanks for your help! YOU DID HELPED ME SO MUCH!
> 
> Nice job. I think converting the regression problem to a classification problem is a possible solution and it is easier to get the statical characteristic from a distribution.

Yeah, this really worked, with a similar performance to faster-rcnn+soft without carefully adjusting parameters on both COCO and medical dataset.

I really need to make an additional comparision with STAC+one stage. I want to use your code to implement STAC+one stage detector, since I'm not familiar with STAC official code and tensorflow(sign). I guess your data process part is the same as STAC and I only need to train twice(firstly superivised and then initiate teacher&student) cancel EMA update, change the loss function and modify the ssod(one stage version)(cancel reliabitity estimate and localization quality estimate) then this will be OK. Anything left out as you can see? 

Thanks a lot and no need to open this issue.

**Comment by Jack-Hu-2001** [comment]
Oh, is it OK to change the sample ratio to make it just select unsup images, such as change it to [0, 1] (in the seconde stage)?
Have you used your code to reimplement STAC+faster-rcnn? If so, you could really teach me a lot on this issue.

**Comment by Jack-Hu-2001** [comment]
This method has worked and after an arxiv article is finished, I'll release the code. Many people have asked me about this from this issue, I reply here.

**Comment by 4-0-4-notfound** [comment]
@Jack-Hu-2001 Hi, when will you release the code, since i'm curious about it~
